Add bank system run by an embeded Jetty server

tsbonev opened the pull request:
Initial commit

Add Jetty server

Add models and DAO interfaces

Add AccountDAO implementation

Add connection test

Add AccountDAO integration tests

Fix account tests

Add UserDAO implementation

Fix users schema name and add creation tests

Add UserDAO tests

Refactor table cleaning in tests

Add transaction table creation test

Add AccountDAO implementation

Add transaction tests

Add service layer

Add HomeServlet

Add index

Add bootstrap

Add login servlet and global alerts

Add register servlet

Fix redirect issues

Add logout functionality

Add user countring functionality

Add login system tests

Add PRG to Login

Add PRG to Logout

Add PRG to Register and adjust tests

Add login filter and transaction functionality

Add transaction history

Refactor packages

Add usernames to history

Remove test logs and clean up minor code

Add authentication and authorization filters

Add global redirect on exception thrown

Add server-side regex validation

Add pagination to DAO

Add page checkers to transaction service

Add pagination to servlet

Adjust js validation to match server-side

Fix authorization checks and forbid looking at foreign accounts

Refactor package names

Rename DAO to repository

Refactor Page to a injected dependency

Add connection per request filter

Refactor db to not include a set account

Remove services

Add transaction on register

Fix jsp href issue

Remove old account servlet

Fix history jsp scope loss

Fix pagination

Add documentation to repositories

Add documentation for filters

Add documentation to helpers

Finish and clean up method documentation

Clean up register servlet

Clean up login servlet

Clean up account servlet

Clean up history servlet

Clean up transaction servlet

Add account servlet test

Fix personal transaction history issue

Add page and transaction system tests

Add transaction repository tests

Added a test for the calculation of balances,
added a test for the proper fetching of the row count
of transactions on a global and a specific userId scope,
added a test for the proper filling of transaction usernames.

Refactor Impl suffix

Capitalize Mysql to MySQL

Add error filter test

Refactor jdbc driver to datasource connection pool

Add 404 ErrorHandler to embeded jetty server

Change account balance test name for better readability

mgenov commented on src/main/java/core/repository/MySQLUserRepository.java:
```diff
@@ -0,0 +1,276 @@
+     *
+     * @param conn that is injected
+     */
+    public void setConnection(Connection conn){
```
Setter injection is more rarely used as it doesn't enforce restrictions, e.g user is not enforced to set the connection. 
It is better to pass connection provider to the constructor. 

```java
public MySqlUserRepository(Provider<Connection> connectionProvider) {
   //...   
} 

 public User getById(int id) {
     // ...
     Connection connection = connectionProvider.get()
  }
```

mgenov commented on src/main/java/core/repository/MySQLUserRepository.java:
```diff
@@ -0,0 +1,276 @@
+
+            ResultSet result = get.executeQuery();
+
+            while (result.next()){
```
The ResultSet iteration is duplicated and in the getById call and everywhere where you are executing select query. 

This duplication could be eliminated by wrapping the iteration into something like Mapper: 

```java
List<User> users = jdbcTemplate.find("select * from userDb", new RowMapper<User>() {
    public User map(ResultSet rs) {
       return new User(rs.getInt("id"));
    }
}
```

mgenov commented on src/main/java/core/repository/MySQLUserRepository.java:
```diff
@@ -0,0 +1,276 @@
+                users.add(user);
+            }
+
+        } catch (SQLException e) {
```
You are not closing the result set ...

mgenov commented on src/main/java/core/repository/MySQLUserRepository.java:
```diff
@@ -0,0 +1,276 @@
+
+        try {
+            PreparedStatement create = conn.prepareStatement(
+                    "CREATE TABLE IF NOT EXISTS userDb(" +
```
userDb doesn't look like a good name of the table.

mgenov commented on src/main/java/core/repository/MySQLUserRepository.java:
```diff
@@ -0,0 +1,276 @@
+    /**
+     * Creates a user table in the database.
+     */
+    public void createUserTable() {
```
This method is not related to the business logic, so it shouldn't be defined here. 

DDL is mostly placed in sql files like: schema/bank.sql. This files could be used for the production set up of the tables and for testing purposes. 

For testing this file could be read as file and the content to be executed as commands to the database.

mgenov commented on src/main/java/core/repository/TransactionRepository.java:
```diff
@@ -0,0 +1,40 @@
+
+    Transaction getById(int id);
+
+    List<Transaction> getAll(int page);
```
getAll of which account ?

mgenov commented on src/main/java/core/repository/TransactionRepository.java:
```diff
@@ -0,0 +1,40 @@
+
+public interface TransactionRepository {
+
+    void setConnection(Connection conn);
```
This method is not related to the interface and cannot be here. I've already point you about the constructor injection.

mgenov commented on src/main/java/core/repository/TransactionRepository.java:
```diff
@@ -0,0 +1,40 @@
+
+    List<Transaction> getAll(int page);
+
+    double getBalance(int userId);
```
getBalance is not related to the transactions

mgenov commented on src/main/java/core/repository/TransactionRepository.java:
```diff
@@ -0,0 +1,40 @@
+
+    double getBalance(int userId);
+
+    void deleteById(int id);
```
Transactions cannot be deleted as they are history.

mgenov commented on src/main/java/core/repository/TransactionRepository.java:
```diff
@@ -0,0 +1,40 @@
+
+    void save(Transaction transaction);
+
+    void update(Transaction transaction);
```
Transactions cannot be updated as they are immutable.

mgenov commented on src/main/java/core/repository/TransactionRepository.java:
```diff
@@ -0,0 +1,40 @@
+
+    List<Transaction> getByOperation(Transaction.Operation operation, int page);
+
+    void fillUsernames(List<Transaction> list);
```
This looks really strange. I suppose that the transaction need to be applied on the creation time. 

If you are going to use names during the retrieval, then you need a TransactionUserView object which to contain transaction information with names of the users that have performed them.

mgenov commented on src/main/java/core/repository/UserRepository.java:
```diff
@@ -0,0 +1,29 @@
+
+    void deleteById(int id);
+
+    void deleteAllRows();
```
not part of the interface.

mgenov commented on src/main/java/core/repository/UserRepository.java:
```diff
@@ -0,0 +1,29 @@
+
+    User getByUsername(String username);
+
+    boolean checkPassword(User user);
```
Could be performed and by the getByUsername in code. Keep in mind that this interface is named Repository, so it's responsible for retrieving of the data and saving, but not for checking of the password. This logic doesn't sound like belongs here.

mgenov commented on src/main/java/core/servlet/filter/ConnectionPerRequest.java:
```diff
@@ -0,0 +1,59 @@
+        }
+        finally {
+            try {
+                ConnectionPerRequest.connection.get().close();
```
You have to ensure that connection is set  in the ThreadLocal as getConnection() may fail in the try block.

mgenov replied:
It's in the finally block, so it will be executed even when connection establishment was failed at first place.

tsbonev replied:
So before attempting to close the connection I should have a check to see if the threadLocal is not null?

mgenov commented on src/main/java/core/servlet/helper/UserCounter.java:
```diff
@@ -0,0 +1,62 @@
+     *
+     * @param username to add
+     */
+    public void addUserToCount(String username){
```
This method call is not safe for concurrent updates, but you are using it for all users of the application ? 

This app could be stateless if you keep the sessions in the database instead of using the Session of the application container. Here is an example database structure which you can use as starting point:

ID | UserID | expiresAt  
------------ | ------------- | ----------
session1 |  1| 2018-09-20 11:30:34 
session2 |  2| 2018-09-20 11:23:18


So when you would like to see the count of active users it will be something like this:
```sql
select count(*) from sessions where expiresAt < now()
```

**A clean up** service may run in background to clear expired sessions as this data will become really huge with the time and will require some retention.

mgenov commented on src/main/java/core/servlet/login/RegisterServlet.java:
```diff
@@ -0,0 +1,163 @@
+    private boolean usernameIsRegistered(String username,
+                                         HttpServletRequest req, HttpServletResponse resp) throws IOException {
+
+        if(userRepository.getByUsername(username).getId() != 0){
```
It is better Optional to be used instead of returning dummy object with 0/""/null values. 

```java
Optional<User> possibleUser = userRepository.getByUsername(username)
if (!possibleUser.isPresent()) {
    page.redirectTo("/register", resp, req, "errorMessage, "Username taken!")
}
```

mgenov commented on src/main/java/core/servlet/transaction/TransactionServlet.java:
```diff
@@ -0,0 +1,223 @@
+@WebServlet("/transaction")
+public class TransactionServlet extends HttpServlet {
+
+    private static String fraction = "^[1-9]{1}[0-9]{0,9}[.,]{1}[0-9]{1,5}$";
```
You can compile them first for performance reasons. 

https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html

mgenov commented on src/test/java/core/repository/TransactionRepositoryTest.java:
```diff
@@ -0,0 +1,261 @@
+
+public class TransactionRepositoryTest {
+
+    MySQLUserRepository userRepository = new MySQLUserRepository();
```
For tests you can create a customer JUnit Rule (DatastoreRule) which to create tables and to erase them after each test. This is going to simplify this logic here.

tsbonev commented:
Removed BL from repositories and their interfaces.
Removed setconnection from repositories.
Refactored ConnectionProvider to work without the method injection, but instead
rely on a constructor injection.
Added a session table into the database.
Added a filter to set a session cookie and
log in the user if a session cookie corresponds to a 
session in the database.
Added a background thread to clean up stale sessions in the database.
Added a separate testing database for the SQL tests.
Added a bank.sql file with the schema for the tables.
Made regex compiled instead of plain strings.
Added mapper methods to counteract duplication in repositories.
Refactored repositories to use Optional instead of sending dummy objects.
Added sql count statement for distinct non-expired sessions and a test for it.

tsbonev commented:
Added a check to the finally statement of the ConnectionPerRequest class.
Refactored the session clean up to use a reentrant lock, moved the clean up sessions
call to before the call to create a new session.

mgenov commented on src/main/java/core/repository/MySQLSessionRepository.java:
```diff
@@ -0,0 +1,203 @@
+                            " WHERE expiresOn > current_timestamp"
+            );
+
+            ResultSet result = select.executeQuery();
```
RS and PS are not closed.

mgenov commented on src/main/java/core/repository/MySQLSessionRepository.java:
```diff
@@ -0,0 +1,203 @@
+            save.setInt(2, session.getUserId());
+            save.setTimestamp(3, session.getExpiresOn());
+
+            save.execute();
```
PS is not closed

mgenov commented on src/main/java/core/repository/MySQLTransactionRepository.java:
```diff
@@ -0,0 +1,279 @@
+@SuppressWarnings("Duplicates")
+public class MySQLTransactionRepository implements TransactionRepository {
+
+    private static int pageSize = 10;
```
It doesn't have to be static as you may have only one instance of the MySqlTransactionRepository.

mgenov commented on src/main/java/core/repository/MySQLTransactionRepository.java:
```diff
@@ -0,0 +1,279 @@
+     *
+     * @param pageSize
+     */
+    public void setPageSize(int pageSize) {
```
Setter injection should be avoided as much as you can. In this case you can pass pageSize as param to the constructor.

mgenov commented on src/main/java/core/repository/MySQLTransactionRepository.java:
```diff
@@ -0,0 +1,279 @@
+
+        List<Transaction> transactions = new ArrayList<>();
+
+        while (rs.next()) {
```
This idea of my mapper proposal was to remove the loop duplication. It's still here in the code.

mgenov commented on src/main/java/core/repository/SessionRepository.java:
```diff
@@ -0,0 +1,16 @@
+
+public interface SessionRepository {
+
+    void save(Session session);
```
registerSession() 
refreshSession() 

sounds more reasonable as names

mgenov commented on src/main/java/core/repository/SessionRepository.java:
```diff
@@ -0,0 +1,16 @@
+public interface SessionRepository {
+
+    void save(Session session);
+    void delete(Session session);
```
sessions.deleteSessionsExpiringAfter(LocalDateTime.now())

mgenov commented on src/main/java/core/repository/TransactionRepository.java:
```diff
@@ -0,0 +1,26 @@
+
+    int getRowsForUserId(int userId);
+
+    void setPageSize(int pageSize);
```
set and get page sizes should not be exposed.

mgenov commented on src/main/java/core/repository/TransactionRepository.java:
```diff
@@ -0,0 +1,26 @@
+
+    Optional<Transaction> getById(int id);
+
+    List<Transaction> getAllGlobally(int page);
```
I'm not sure that you need such method.

mgenov commented on src/main/java/core/servlet/filter/SessionFilter.java:
```diff
@@ -0,0 +1,143 @@
+    /**
+     * Starts a thread that cleans up stale sessions.
+     */
+    private void runCleanUp(){
```
This looks overwhelming. You may create a SessionCleanUpTask which to executed regularly in background for the whole task to clean user sessions. 

Guava services are really good way for dealing with this stuff as they provide start/stop features: 
https://github.com/google/guava/wiki/ServiceExplained

mgenov commented on src/main/java/core/servlet/filter/SessionFilter.java:
```diff
@@ -0,0 +1,143 @@
+            return;
+        }
+
+        LoginSession loginSession = new LoginSession();
```
If you are using persistence, then these sessions are no longer required. Also session need to start in the LoginServlet I suppose as it's the authentication part of the system.

tsbonev replied:
These session in the sense that I needn't pass the HttpSession a specific object with data or that I needn't use a HttpSession?

mgenov replied:
That you need HttpSessoin. If you are building stateless app it should be avoided.

tsbonev replied:
So without a HttpSession I should keep an authenticated session object in the server, specific to the thread, and retrieve the user information from there, destroying it when the user logs out, or should I just keep the cookie's session id and check the persistence session on every request?

mgenov replied:
Yes, the session need to be retrieved from the server on each request.

tsbonev replied:
Should every successful retrieval of the session from the database make a new cookie and update the token of the persistent session?

mgenov replied:
The session expiration time need to be updated in the persistence, but adding of a new cookie is not required. 

You may put lifetime of the cookie to be 2 years, so it would not require update.

tsbonev replied:
In such a case, if the log out doesn't clear the cookie, how would it prevent the filter from just logging back into with the same account?
And if the log out should clear the cookie then shouldn't I remove the associated session instead of leaving it be orphaned until the next clean up task?

mgenov replied:
The cookie is just a link between browser and the session of the server. If you remove the server session, then this link is going to be removed.

mgenov commented on src/main/java/core/servlet/filter/util/SessionCleanup.java:
```diff
@@ -0,0 +1,46 @@
+    @Override
+    public void run() {
+
+        if(SessionCleanup.lock.tryLock()){
```
The locking is happening in the database now, so no locking in the code is required.

mgenov commented on src/main/java/core/servlet/filter/util/SessionCleanup.java:
```diff
@@ -0,0 +1,46 @@
+            try {
+                ConnectionProvider provider = new MySQLConnectionProvider();
+                SessionRepository sessionRepo = new MySQLSessionRepository(provider);
+                List<Session> list = sessionRepo.getAll();
```
You can use query syntax with delete to clean up the expired sessions. 

```java
int clearedSessions = sessionRepository.cleanSessionsExpiringAfter(Instant.now())
```

mgenov commented on src/main/java/core/servlet/helper/SIDHandler.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.Arrays;
+import java.util.Optional;
+
+public class SIDHandler {
```
This code doesn't look really useful. The Cookie retrievel is happening only in the filter whether the setting of the cookie should happen either in the filter or in the login page. So there are 2 places for such usage.

mgenov commented on src/main/java/core/servlet/helper/UserCounter.java:
```diff
@@ -0,0 +1,20 @@
+    }
+
+    public int getCount(){
+        SessionRepository sessionRepo = new MySQLSessionRepository(provider);
```
You can pass the SessionRepository directly to the constructor instead of constructing it in the code. This is going to decouple UserCounter from the MySQL Related stuff.

mgenov commented on src/main/java/core/servlet/login/RegisterServlet.java:
```diff
@@ -0,0 +1,178 @@
+     * @return validation result
+     * @throws IOException
+     */
+    private boolean validateForm(String username, String password,
```
validateLoginCredentials sounds more reasonable as name

mgenov commented on src/main/java/core/servlet/login/RegisterServlet.java:
```diff
@@ -0,0 +1,178 @@
+        user.setUsername(username);
+        user.setPassword(password);
+
+        if(!usernameIsRegistered(username, req, resp)){
```
Could be improved with something like:

```java
try {
  userRepository.registerIfNotExists(user)
} catch(UserAlreadyExistsException e) {
   page.redirectTo("/login", resp, req, "errorMessage", "User Already Exists")
}
```

mgenov commented on src/main/java/core/servlet/login/RegisterServlet.java:
```diff
@@ -0,0 +1,178 @@
+     *
+     * @param user with filled id
+     */
+    private boolean trySetUserId(User user, HttpServletRequest req, HttpServletResponse resp) throws IOException {
```
This looks really strange. I'm not sure that you this at all. 

```java
User registeredUser = userRepository.registerUserIfNotExists(new UserRegistrationRequest(username, password));
// registerUser.getId() should be visible ... 
```

Keep in mind that it's good registration request to be different from the regular User domain object.

mgenov commented on src/main/java/core/servlet/transaction/AccountServlet.java:
```diff
@@ -0,0 +1,135 @@
+     */
+    private Optional<User> getUser(HttpServletRequest req){
+
+        SessionHandler session = (SessionHandler) req.getSession().getAttribute("authorized");
```
This user should be retrieved only from the Session.

mgenov commented on src/test/java/core/repository/SessionRepositoryTest.java:
```diff
@@ -0,0 +1,80 @@
+
+    @Before
+    public void setUp(){
+        ConnectionPerRequest.connection.set(creator.getConnection());
```
This is not required as you have a testing ConnectionProvider with to accept the estalished connection. Now you are using production filter related code with static initializes which makes test ugly.

mgenov commented on src/test/java/core/repository/TransactionRepositoryTest.java:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    @Test
+    public void getRowsForUser(){
```
fewTransactions()

mgenov commented on src/test/java/core/repository/TransactionRepositoryTest.java:
```diff
@@ -0,0 +1,147 @@
+
+        transactionRepository.save(transactionTwo);
+
+        assertThat(transactionRepository.getRowsForUserId(0), is(2));
```
getUserTransactions instead of getRowsForUser

mgenov commented on src/test/java/core/repository/TransactionRepositoryTest.java:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    @Test
+    public void getAllRowsIfIdIsZero(){
```
Doesn't look like e reasonable behaviour.

mgenov commented on src/test/java/core/repository/TransactionRepositoryTest.java:
```diff
@@ -0,0 +1,147 @@
+
+        transactionRepository.save(transaction);
+
+        List<Transaction> dbTransactions = transactionRepository.getAllByUserId(transaction.getUserId(), 1);
```
It's not clear whether this test verifies that filter is applied or you are testing the case when only one transaction is available. 

```java
List<Transaction> transactions = getUserTransactions(transaction.getUserId(), 10) 
assertThat(transactions, is(equalto(Collections.singleton(new Transaction()))
```

mgenov commented on src/test/java/core/repository/TransactionRepositoryTest.java:
```diff
@@ -0,0 +1,147 @@
+        transaction.setAmount(200.0d);
+        transaction.setOperation(Transaction.Operation.DEPOSIT);
+        java.util.Date utilDate = Date.from(Instant.now());
+        java.sql.Date sqlDate = new Date(utilDate.getTime());
```
sql.Date should not be used in the test I suppose.

mgenov commented on src/test/java/core/repository/util/CreateTestTable.java:
```diff
@@ -0,0 +1,72 @@
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+
+public class CreateTestTable {
```
CreateTestTable and DropTestTable could be combined in a single object: TestDatabase 

it may have a common method: cleanUpTables()  which erases all tables. It's not required each test to know which production tables are used by the production code.

tsbonev commented:
Should I make a JdbcTemplate for the inserts, updates and single selects or should I just use it for the list returns?

mgenov commented:
Yes, for inserts and updates will simplify your code in the repository and will remove duplication.

tsbonev commented:
Refactored packages to fit orthogonal design.
Added JdbcTemplates to remove duplication from repos.
Cleaned up transaction repository behaviours and unused methods.
Removed view for global transactions.
Removed admin access to foreign accounts.
Removed get-based account and history viewing.
Removed row counting.
Moved pageSize to the servlet's constructor.
Combined test table creation and dropping into a single object.
Cleaned up static filter usage in tests.
Added a register request instead of directly saving a user domain object.
Removed HttpSessions and LoginSession objects, and implemented persistence sessions.
Refactored user getting to only retrieve the user attached to the session.

tsbonev commented:
Removed unused getAll from SessionRepository and cleaned up some method documentations.

tsbonev commented:
Removed the ThreadLocal set from the login servlet and left it to be the responsibility of the session filter.

mgenov commented on src/main/java/com/clouway/bankapp/adapter/mysql/MySQLConnectionCreator.java:
```diff
@@ -0,0 +1,45 @@
+import java.sql.Connection;
+import java.sql.SQLException;
+
+public class MySQLConnectionCreator implements ConnectionCreator {
```
Doesn't look reasonable. The creator should return data source which to be used by the Connection Provider. Now client code is able to create multiple data sources and using one connection of them.

tsbonev replied:
So the provider should call the creator once, get the datasource, and then give out connections from it? In such a case what role would the ConnectionPerRequest have, as right now the provider returns the ThreadLocal connection that the CPR filter opens via the creator.

mgenov replied:
In short: The filter should have a reference to the DataSource and have to take connection from it when the request starts.

mgenov commented on src/main/java/com/clouway/bankapp/adapter/mysql/MySQLConnectionCreator.java:
```diff
@@ -0,0 +1,45 @@
+    public Connection get() {
+
+        try {
+            String url = "jdbc:mysql://localhost:3306/bank?useSSL=false" +
```
keeping credentials in code is not a good idea.

mgenov replied:
You can use configuration files, command line arguments or environment variables about this. 

https://12factor.net/config

mgenov commented on src/main/java/com/clouway/bankapp/adapter/web/filter/SessionFilter.java:
```diff
@@ -0,0 +1,118 @@
+     */
+    private Session setUpSession(Optional<Session> optionalSession){
+
+        Session retrievedSession = optionalSession.get();
```
Doesn't look reasonable optional to be passed as param and isPresent() to not be used in the method.

mgenov commented on src/main/java/com/clouway/bankapp/adapter/web/filter/SessionFilter.java:
```diff
@@ -0,0 +1,118 @@
+                Cookie SID = SessionCookie.get();
+                req.setAttribute("SID", SID.getValue());
+
+                Optional<Session> optionalSession = sessionRepo.getBySessionId(SID.getValue());
```
What if session is expired at that time  ?

// Lookup for session which should be available at specified time. 
sessionRepository.getSessionAvailableAt(SID.getValue(), Instant.now())

mgenov replied:
possibleSession sounds better as name.

mgenov commented on src/main/java/com/clouway/bankapp/adapter/web/filter/SessionFilter.java:
```diff
@@ -0,0 +1,118 @@
+
+                    Session retrievedSession = setUpSession(optionalSession);
+
+                    sessionRepo.refreshSession(retrievedSession);
```
Maybe something like this is more appropriate:

```java
// Needs to be ensured that the /login page is not the target one
if (!possibleSession.isPresent()) {
   resp.sendRedirect("/login")
}

Session session = possibleSession.get()
User user = userRepository.getUserById(session.userId)

// you can use thread local and for the signed user and UserProvider for retrieving 
// it from code
userContext.set(user)
```

tsbonev replied:
I refactored the filter to add the user to a separate userContext thread local but haven't implemented the redirection if a session isn't present as the authentication filter already denies access to pages that require a session.

mgenov replied:
As I understand your point, you have 2 filters which are doing similar things.

mgenov commented on src/main/java/com/clouway/bankapp/adapter/web/filter/util/SessionCleanupTask.java:
```diff
@@ -0,0 +1,54 @@
+    @Override
+    public void startUp(){
+
+        ConnectionCreator creator = new MySQLConnectionCreator();
```
This code doesn't look reasonable to be part of the Task. The Task may accept SessionRepository interface which to use. Now this service knows for the JDBC package, for the connection provider, for the templates and for the MySQLSessionRepository.

tsbonev replied:
Currently I start the clean up task before I start the Jetty server so a separate creator, provider and template are required by the session repo to run. Should I override the sessionRepo interface to make it work without the connection provider or should I move the cleanup task into the jetty server and make it run the ConnectionPerRequest filter before it runs an iteration?

mgenov replied:
You have to move common dependencies one level up to the call stack.

mgenov commented on src/main/java/com/clouway/bankapp/adapter/web/transaction/HistoryServlet.java:
```diff
@@ -0,0 +1,149 @@
+     * @param req servlet request
+     * @param transactions the list of transactions
+     */
+    private void setUpPagination(int currPage,
```
The code in this method reads the transactions using the provided arguments. Doesn't look like a set up of pagination.

mgenov commented on src/main/java/com/clouway/bankapp/adapter/web/transaction/HistoryServlet.java:
```diff
@@ -0,0 +1,149 @@
+            transactions.remove(transactions.size() - 1);
+        }
+
+        List<TransactionUserView> transactionUserViews = fillUsernames(transactions);
```
This should happen by the persistence layer as he is able to join the 2 tables. It's a bad habbit data to be prepended at some time to an object as it's evaluates the immutability principle. Sometimes it's required, but there are rare cases.

mgenov commented on src/main/java/com/clouway/bankapp/adapter/web/transaction/TransactionServlet.java:
```diff
@@ -0,0 +1,198 @@
+
+    private static Pattern fractionPattern = Pattern.compile("^[1-9]{1}[0-9]{0,9}[.,]{1}[0-9]{1,5}$");
+    private static Pattern wholeNumberPattern = Pattern.compile("^[1-9]{1}[0-9]{0,9}$");
+    private static Pattern leadingZeroFractionPattern = Pattern.compile("^[0]{1}[.,]{1}[0-9]{0,4}[1-9]{1}$");
```
amountPattern sounds better to me.

mgenov commented on src/main/java/com/clouway/bankapp/adapter/web/transaction/TransactionServlet.java:
```diff
@@ -0,0 +1,198 @@
+
+        User user = getSession().getUser();
+
+        Transaction transaction = setUpTransaction(amount, operation, user);
```
```java
Transaction registeredTransaction = transactionRepository.registerTransaction(user, amount, operation)
```

mgenov commented on src/main/java/com/clouway/bankapp/adapter/web/transaction/TransactionServlet.java:
```diff
@@ -0,0 +1,198 @@
+     */
+    private Transaction setUpTransaction(double amount, Operation operation, User user){
+
+        Transaction transaction = new Transaction();
```
It is better Transaction to accept all of these arguments in the constructor, instead as setters.

mgenov commented on src/main/java/com/clouway/bankapp/adapter/web/transaction/TransactionServlet.java:
```diff
@@ -0,0 +1,198 @@
+     */
+    private boolean validateAmountFormat(String amount,
+                                         HttpServletRequest req, HttpServletResponse resp) throws IOException {
+        if(!matchToPattern(amount, wholeNumberPattern)
```
You may validate it with a single expression.

tsbonev commented:
Moved credentials into a config file.
Refactored ConnectionCreator to return a datasource and create one if it hasn't already.
Refactored ConnectionProvider to get a connection from the creator if the filter hasn't already  set one.
Merged the JdbcTemplates into one class, made them use varargs instead of a map.
Fixed variable and method naming issues, retouched some method documentations.
Moved the authentication filter responsibility into the session filter.
Refactored session getting to include a timestamp with which to judge.
Unbound user from session and created a new threadlocal to contain the user.
Moved common dependencies one level up and passed them to the jetty server via constructor.
Refactored sessionCleanUp to not instantiate separate dependencies.
Moved the filling usernames responsibility to the repository with a join sql query.
Merged transaction regex patterns into one.
Refactored transaction creation in the servlet to use the constructor instead of setters.

mgenov commented on src/main/java/com/clouway/bankapp/adapter/mysql/MySQLConnectionCreator.java:
```diff
@@ -0,0 +1,86 @@
+
+        try{
+
+            config = new File("src/main/resources/config/connection.config");
```
Doesn't look reasonable as in production build src/main/resources won't be available. Also it's not a good idea configuration files to be packaged with jars.

tsbonev replied:
Moved configuration file into the root of the classpath.

mgenov commented on src/main/java/com/clouway/bankapp/adapter/mysql/MySQLConnectionCreator.java:
```diff
@@ -0,0 +1,86 @@
+     * @param field to read
+     * @return found value
+     */
+    private String readConfig(String field){
```
The configuration or part of it may be passed to the connection creator, but not to let it load it himself. This means that this class has 2 roles and 2 reasons to change.

tsbonev replied:
Refactored config file to be a .properties and read it from the root of the classpath, passing it to the ConnectionCreator via constructor.

mgenov commented on src/main/java/com/clouway/bankapp/adapter/mysql/MySQLConnectionProvider.java:
```diff
@@ -0,0 +1,41 @@
+    public Connection get() {
+
+        try {
+            if(ConnectionPerRequest.connection.get() == null){
```
This logic should be in the ConnectionPerRequest**Filter** instead of exposing internalls to the callers. 

```java
return ConnectionPerRequestFilter.get() 
```

tsbonev replied:
Added a get() method to the CPRF and put the responsibility of checking the connection there.

mgenov commented on src/main/java/com/clouway/bankapp/adapter/mysql/MySQLSessionRepository.java:
```diff
@@ -0,0 +1,129 @@
+     * Maps the rows from a result session
+     * to a session object.
+     */
+    private RowMapper<Session> rowMapper = rs -> new Session(
```
sessionRowMapper as it's not a generic one.

tsbonev replied:
Renamed variable.

mgenov commented on src/main/java/com/clouway/bankapp/adapter/mysql/MySQLSessionRepository.java:
```diff
@@ -0,0 +1,129 @@
+     *
+     * @param sessionId
+     */
+    public void logSessionOut(String sessionId) {
```
terminateSession

tsbonev replied:
Renamed variable

mgenov commented on src/main/java/com/clouway/bankapp/adapter/mysql/MySQLSessionRepository.java:
```diff
@@ -0,0 +1,129 @@
+     * @param timestamp timestamp to compare to
+     * @return found session
+     */
+    public Optional<Session> getBySessionAvailableAt(String sessionId, Timestamp timestamp) {
```
getSessionAvailableAt

tsbonev replied:
Renamed variable.

mgenov commented on src/main/java/com/clouway/bankapp/adapter/mysql/MySQLTransactionRepository.java:
```diff
@@ -0,0 +1,82 @@
+     * @param page to return
+     * @return transactions in a given page
+     */
+    public List<Transaction> getUserTransactionsPaginated(int id, int page, int pageSize) {
```
You can use overloading here as adding of "paginated" and "not paginated" stuff in the name doesn't sound good. The Repository doesn't have to know anything about pages. It has to know how many items are in there, give me next items, give me previous items and etc. 

List<Transaction> getUserTransactions(int userId, int page, int pageSize) 
List<Transaction> getUserTransactions(int userId)

tsbonev replied:
Overloaded the method .

mgenov commented on src/main/java/com/clouway/bankapp/adapter/mysql/MySQLUserRepository.java:
```diff
@@ -0,0 +1,142 @@
+     * Maps the rows from a result session
+     * to a user object.
+     */
+    private RowMapper<User> rowMapper = rs -> new User(
```
userRowMapper

tsbonev replied:
Renamed variable.

mgenov commented on src/main/java/com/clouway/bankapp/adapter/web/login/RegisterServlet.java:
```diff
@@ -0,0 +1,131 @@
+     */
+    private void createInitialTransaction(User user){
+
+        Transaction transaction = new Transaction();
```
Constructor could be used instead of setters.

tsbonev replied:
Used constructor to create a transaction.

tsbonev commented:
PR is up to date with requested changes.

mgenov commented on src/main/java/server/Jetty.java:
```diff
@@ -0,0 +1,167 @@
+
+
+        try {
+            if(!testDatabaseConnection()){
```
DB connection check should be performed in the entry point.

tsbonev replied:
Moved db test to entry point

tsbonev commented:
Changes are up to date.

tsbonev commented:
Moved cookie adding to session filter as accessing the site from a browser with no cookie was impossible.
